app/src/main/java/com/x8bit/bitwarden/data/platform/manager/restriction/RestrictionManagerImpl.kt

```diff
@@ -0,0 +1,137 @@
+package com.x8bit.bitwarden.data.platform.manager.restriction
+
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.content.RestrictionsManager
+import android.os.Bundle
+import com.x8bit.bitwarden.data.platform.manager.AppForegroundManager
+import com.x8bit.bitwarden.data.platform.manager.dispatcher.DispatcherManager
+import com.x8bit.bitwarden.data.platform.manager.model.AppForegroundState
+import com.x8bit.bitwarden.data.platform.repository.EnvironmentRepository
+import com.x8bit.bitwarden.data.platform.repository.model.Environment
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+
+/**
+ * The default implementation of the [RestrictionManager].
+ */
+class RestrictionManagerImpl(
+    appForegroundManager: AppForegroundManager,
+    dispatcherManager: DispatcherManager,
+    private val context: Context,
+    private val environmentRepository: EnvironmentRepository,
+    private val restrictionsManager: RestrictionsManager,
+) : RestrictionManager {
+    private val ioScope = CoroutineScope(dispatcherManager.io)
+    private val intentFilter = IntentFilter(Intent.ACTION_APPLICATION_RESTRICTIONS_CHANGED)
+    private val restrictionsChangedReceiver = RestrictionsChangedReceiver()
+    private var isReceiverRegistered = false
+
+    init {
+        appForegroundManager
+            .appForegroundStateFlow
+            .onEach {
+                when (it) {
+                    AppForegroundState.BACKGROUNDED -> handleBackground()
+                    AppForegroundState.FOREGROUNDED -> handleForeground()
+                }
+            }
+            .launchIn(ioScope)
```

Comment:
Why `ioScope`? Given that you are registering receivers I would think you might actually want this to be based on the main dispatcher.

Reply:
That is a good point!